test/include/test/bench.h

```diff
@@ -58,3 +58,14 @@ compare_funcs(uint64_t nwarmup, uint64_t niter, const char *name_a,
 
 	dallocx(p, 0);
 }
+
+static void *
```

Comment:
Let's do `static inline`; we want to avoid an actual function call.

Curious did the generated asm look any different in your case?

Reply:
I am checking the assembly, it's a bit weird that for sometimes calling `no_opt_ptr` even makes the compiler optimize away the `expect_ptr_not_null`.

New version:
```
000000000040e0e0 <_ZL10new_deletev>:
new_delete(void) {
  40e0e0:       48 83 ec 08             sub    $0x8,%rsp
        void* p = ::operator new(1);
  40e0e4:       bf 01 00 00 00          mov    $0x1,%edi
  40e0e9:       e8 52 36 ff ff          call   401740 <_Znwm@plt>
  40e0ee:       48 89 c7                mov    %rax,%rdi
}
  40e0f1:       48 83 c4 08             add    $0x8,%rsp
        ::operator delete((void *)p);
  40e0f5:       e9 d6 36 ff ff          jmp    4017d0 <_ZdlPv@plt>
  40e0fa:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)
```

Original version:
```
000000000040e200 <_ZL10new_deletev>:
new_delete(void) {
  40e200:       48 81 ec 18 02 00 00    sub    $0x218,%rsp
        void* volatile p = ::operator new(1);
  40e207:       bf 01 00 00 00          mov    $0x1,%edi
  40e20c:       e8 2f 35 ff ff          call   401740 <_Znwm@plt>
  40e211:       48 89 44 24 08          mov    %rax,0x8(%rsp)
        expect_ptr_not_null((void *)p, "Unexpected new failure");
  40e216:       48 8b 44 24 08          mov    0x8(%rsp),%rax
  40e21b:       48 85 c0                test   %rax,%rax
  40e21e:       74 18                   je     40e238 <_ZL10new_deletev+0x38>
        ::operator delete((void *)p);
  40e220:       48 8b 7c 24 08          mov    0x8(%rsp),%rdi
  40e225:       e8 a6 35 ff ff          call   4017d0 <_ZdlPv@plt>
}
  40e22a:       48 81 c4 18 02 00 00    add    $0x218,%rsp
  40e231:       c3                      ret
  40e232:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)
        expect_ptr_not_null((void *)p, "Unexpected new failure");
  40e238:       6a 00                   push   $0x0
  40e23a:       41 b9 0e 00 00 00       mov    $0xe,%r9d
  40e240:       41 b8 30 b1 48 00       mov    $0x48b130,%r8d
  40e246:       b9 30 b3 48 00          mov    $0x48b330,%ecx
  40e24b:       6a 00                   push   $0x0
  40e24d:       ba 50 b1 48 00          mov    $0x48b150,%edx
  40e252:       be 00 01 00 00          mov    $0x100,%esi
  40e257:       68 08 b2 48 00          push   $0x48b208
  40e25c:       68 0f b2 48 00          push   $0x48b20f
  40e261:       48 8d 7c 24 30          lea    0x30(%rsp),%rdi
  40e266:       e8 70 44 ff ff          call   4026db <jet_malloc_snprintf>
  40e26b:       48 83 c4 20             add    $0x20,%rsp
  40e26f:       ba 33 b2 48 00          mov    $0x48b233,%edx
  40e274:       31 c0                   xor    %eax,%eax
  40e276:       48 8d bc 24 10 01 00    lea    0x110(%rsp),%rdi
  40e27d:       00 
  40e27e:       be 00 01 00 00          mov    $0x100,%esi
  40e283:       e8 53 44 ff ff          call   4026db <jet_malloc_snprintf>
  40e288:       48 8d b4 24 10 01 00    lea    0x110(%rsp),%rsi
  40e28f:       00 
  40e290:       48 8d 7c 24 10          lea    0x10(%rsp),%rdi
  40e295:       e8 f6 c7 07 00          call   48aa90 <p_test_fail>
  40e29a:       eb 84                   jmp    40e220 <_ZL10new_deletev+0x20>
  40e29c:       0f 1f 40 00             nopl   0x0(%rax)
```

Tried to add `inline` to `no_opt_ptr`, but the assembly barely changes. Not sure why, but this may indicate that using `volatile asm` may not solve problems like issue #2356 ?

Reply:
It actually shows the benefit of using `asm volatile` -- the default operator new can't return NULL (instead it will throw std::bad_alloc). If we mark the pointer as volatile, the compiler can't optimize away the NULL check since it can't assume the value doesn't change. With the less scoped asm volatile, it knows at the point of check the ptr can't be NULL so it goes away.